packages/world/src/utils/InventoryUtils.sol

```diff
@@ -166,94 +231,93 @@ library InventoryUtils {
     require(stackable > 0, "Object type cannot be added to inventory");
 
+    // Check slot is within bounds for this entity type
+    uint16 maxSlots = owner.getObjectType().getMaxInventorySlots();
+    require(maxSlots > 0, "Invalid slot");
+    require(slot < maxSlots, "Slot exceeds entity's max inventory");
+
     InventorySlotData memory slotData = InventorySlot._get(owner, slot);
 
     if (slotData.objectType.isNull()) {
-      _useEmptySlot(owner, slot);
-      InventorySlot._setObjectType(owner, slot, objectType);
-      _addToTypeSlots(owner, objectType, slot);
+      setBit(owner, slot);
+      InventorySlot._set(owner, slot, EntityId.wrap(0), objectType, amount);
     } else {
       require(slotData.objectType == objectType, "Cannot store different object types in the same slot");
+      uint16 newAmount = slotData.amount + amount;
+      require(newAmount <= stackable, "Object does not fit in slot");
+      InventorySlot._setAmount(owner, slot, newAmount);
     }
-
-    uint16 newAmount = slotData.amount + amount;
-    require(newAmount <= stackable, "Object does not fit in slot");
-
-    InventorySlot._setAmount(owner, slot, newAmount);
   }
 
-  // IMPORTANT: this does not burn tool ores
   function removeEntity(EntityId owner, EntityId entity) internal {
-    uint16[] memory slots = Inventory._getOccupiedSlots(owner);
-    for (uint256 i = 0; i < slots.length; i++) {
-      if (entity == InventorySlot._getEntityId(owner, slots[i])) {
-        _recycleSlot(owner, slots[i]);
-        Mass._deleteRecord(entity);
-        return;
-      }
-    }
-
-    revert("Entity not found");
+    uint16 slot = findEntity(owner, entity);
+    clearBit(owner, slot);
+    InventorySlot._deleteRecord(owner, slot);
+    Mass._deleteRecord(entity);
   }
 
   function removeEntityFromSlot(EntityId owner, uint16 slot) internal {
     EntityId entity = InventorySlot._getEntityId(owner, slot);
     require(entity.exists(), "Not an entity");
 
-    _recycleSlot(owner, slot);
+    clearBit(owner, slot);
+    InventorySlot._deleteRecord(owner, slot);
     Mass._deleteRecord(entity);
   }
 
   function removeObject(EntityId owner, ObjectType objectType, uint16 amount) internal {
     require(amount > 0, "Amount must be greater than 0");
     require(!objectType.isNull(), "Empty slot");
 
-    // Check if there are any slots with this object type
-    uint256 numTypeSlots = InventoryTypeSlots._length(owner, objectType);
-    require(numTypeSlots > 0, "Not enough objects of this type in inventory");
-
     uint16 remaining = amount;
+    uint256[] memory bitmap = InventoryBitmap._getBitmap(owner);
 
-    // Iterate from end to minimize array shifts
-    for (uint256 i = numTypeSlots; i > 0 && remaining > 0; i--) {
-      uint16 slot = InventoryTypeSlots._getItem(owner, objectType, i - 1);
-      uint16 currentAmount = InventorySlot._getAmount(owner, slot);
+    // walk backwards so clearing whole trailing words is cheaper
```

Comment:
why is it backwards here but not in `addObject`

Reply:
because we prune the trailing words if they are empty